### PR TITLE
Add location, attribute visitors to Tast_iterator/Tast_mapper

### PR DESCRIPTION
**Pull request by dyzsr**
Add `location` visitor to modules Tast_iterator and Tast_mapper, then operations like removing locations can be implemented just as Ast_iterator and Ast_mapper.

### CODE REVIEW

**Comment by gasche** [comment]
I understand the use-case and I think this would be an interesting addition. Two remarks:

- Currently trunk development is frozen during the Multicore OCaml merge, so this contribution may have to wait to be integrated even if there is consensus in favor.

- It's hard to tell whether locations have been forgotten. It would be possible to have a reasonably-good test for this by writing a compiler-libs program that takes an input program, parses and type-checks it (`driver/compile_common.mli` would help for this), erases all locations using Tast_mapper, and then prints the program using the `-dtypedtree` printer. Then we can easily grep the output for the input filename, and catch potentially missing locations. Three sub-remarks:
  + This hypothetical test relies on an input files that exercises all or most components of the typedtree; I don't know if one exists, but it could be useful for other things as well. But if one adds new typedtree features, it's easy to forget to update the test file.
  + The test could miss an un-erased location if it is also ignored by the pretty-printer. In particular, if we add more locations to existing typedtree nodes after the fact, we are likely to forget to update the typedtree printer (just like tast_mapper), and the test would not catch the issue.
  + Given the two limitations above, I'm not convinced anymore that this hypothetical test is worth it, and I would not suggest trying to add it to the PR. I kept the discussion in case it is of interest to someone.

File: typing/tast_iterator.ml [hunk 1]
```diff
@@ -99,20 +117,30 @@ let structure_item sub {str_desc; str_env; _} =
   | Tstr_class list ->
       List.iter (fun (cls,_) -> sub.class_declaration sub cls) list
   | Tstr_class_type list ->
-      List.iter (fun (_, _, cltd) -> sub.class_type_declaration sub cltd) list
-  | Tstr_include incl -> include_infos (sub.module_expr sub) incl
+      List.iter (fun (_, s, cltd) ->
+        iter_loc sub s; sub.class_type_declaration sub cltd) list
+  | Tstr_include incl -> include_infos sub (sub.module_expr sub) incl
   | Tstr_open od -> sub.open_declaration sub od
   | Tstr_attribute _ -> ()
```
Comment (nojb) [inline typing/tast_iterator.ml]: Note that `attribute` and `attributes` types contain `Location.t` values in them. These need to be added to the traversal code.

**Review by nojb — APPROVED** [review]
LGTM (modulo a small point, see comments). I checked the changes in `Tast_iterator` in detail to make sure no location had been forgotten; I only skimmed the ones in `Tast_mapper`, as I assume they are similar.

@dyzsr: can you please include locations appearing in attributes, and rebase the patch on `trunk`? Thanks!

For testing, I don't think we test the similar code in `Ast_iterator` and `Ast_mapper`, so am inclined to accept this contribution without a test.

**Comment by dyzsr** [comment]
PTAL @nojb @gasche

**Pull request merged**